### PR TITLE
V0.11.0+core precompile

### PR DESCRIPTION
**Pull request by AgwaB**

### CODE REVIEW

File: x/evm/statedb/statedb.go [hunk 1]
```diff
@@ -455,9 +498,33 @@ func (s *StateDB) Commit() error {
...
 				}
-				s.keeper.SetState(s.ctx, obj.Address(), key, value.Bytes())
+				s.keeper.SetState(ctx, obj.Address(), key, value.Bytes())
 			}
 		}
 	}
 	return nil
 }
+
+func (s *StateDB) branchNewCacheCtx() *cacheCtx {
```
Comment (junbeomlee) [inline x/evm/statedb/statedb.go]: how about `branchNewCacheCtx` -> `newCacheCtx`

File: x/evm/statedb/statedb.go [hunk 2]
```diff
@@ -29,8 +34,9 @@ var _ vm.StateDB = &StateDB{}
 // * Contracts
 // * Accounts
 type StateDB struct {
-	keeper Keeper
-	ctx    sdk.Context
+	keeper   Keeper
+	ctx      sdk.Context
+	cacheCtx []*cacheCtx // It must be the same size as validRevisions.
```
Comment (junbeomlee) [inline x/evm/statedb/statedb.go]: how about `cacheCtx` -> `cacheCtxs`

File: x/evm/statedb/statedb.go [hunk 3]
```diff
@@ -414,9 +420,32 @@ func (s *StateDB) Snapshot() int {
...
+	ctx := s.ctx
+	cacheCtx := s.latestCacheCtx()
+	if cacheCtx != nil {
+		ctx = cacheCtx.ctx
+	}
+	return ctx
+}
+
+func (s *StateDB) Flush() error {
+	ctx := s.ctx
```
Comment (junbeomlee) [inline x/evm/statedb/statedb.go]: this method can be
```
func (s *StateDB) Flush() error {
        ctx := s.getLatestCtx()
	return return s.commit(ctx)
}
```

File: x/evm/statedb/statedb.go [hunk 4]
```diff
@@ -414,9 +420,32 @@ func (s *StateDB) Snapshot() int {
...
 	s.validRevisions = append(s.validRevisions, revision{id, s.journal.length()})
+	s.cacheCtx = append(s.cacheCtx, s.branchNewCacheCtx())
 	return id
 }
 
+func (s *StateDB) GetLatestCtx() interface{} {
+	return s.getLatestCtx()
+}
+
+func (s *StateDB) getLatestCtx() sdk.Context {
```
Comment (junbeomlee) [inline x/evm/statedb/statedb.go:431]: this method can be
```
func (s *StateDB) getLatestCtx() sdk.Context {
	if len(s.cacheCtx) == 0 {
		return s.ctx
	}

	return s.cacheCtx[len(s.cacheCtx)-1]
}
```
and can remove `latestCacheCtx()` method

File: x/evm/statedb/statedb.go (hunk 1 above)
Comment (AgwaB) [reply x/evm/statedb/statedb.go]: Good!

File: x/evm/statedb/statedb.go (hunk 3 above)
Comment (AgwaB) [reply x/evm/statedb/statedb.go]: Good

File: x/evm/statedb/statedb.go [hunk 5]
```diff
@@ -431,22 +455,36 @@ func (s *StateDB) RevertToSnapshot(revid int) {
...
 }
 
 // Commit writes the dirty states to keeper
 // the StateDB object should be discarded after committed.
 func (s *StateDB) Commit() error {
+	if s.nextRevisionID == 0 {
+		return s.commit(s.ctx)
+	}
+
+	for idx := range s.cacheCtxs {
```
Comment (hea9549) [inline x/evm/statedb/statedb.go]: how about `s.cacheCtx` -> `len(s.cacheCtx)`?
I think len function is more clear index expression

File: x/evm/statedb/statedb.go [hunk 6]
```diff
@@ -29,8 +34,9 @@ var _ vm.StateDB = &StateDB{}
 // * Contracts
 // * Accounts
 type StateDB struct {
-	keeper Keeper
-	ctx    sdk.Context
+	keeper    Keeper
+	ctx       sdk.Context
```
Comment (hea9549) [inline x/evm/statedb/statedb.go:38]: ctx is genesis ctx?

**Review by junbeomlee — APPROVED** [review]

File: x/evm/statedb/statedb.go (hunk 6 above)
Comment (AgwaB) [reply x/evm/statedb/statedb.go:38]: yes

File: x/evm/keeper/statedb_test.go [hunk 7]
```diff
@@ -792,3 +792,219 @@ func (suite *KeeperTestSuite) _TestForEachStorage() {
...
+				amounts[0] + amounts[1],
+				amounts[0] + amounts[1],
+				amounts[0] + amounts[1],
+				amounts[0] + amounts[1],
+			},
+		},
+	}
+
+	for _, tc := range testCase {
+		suite.Run(tc.name, func() {
```
Comment (zeroFruit) [inline x/evm/keeper/statedb_test.go:971]: - All core test cases has the test name of `fmt.Sprintf("Case %s", tc.name)`
- Is ethermint testcase name convention is `tc.name` ?

File: x/evm/keeper/statedb_test.go [hunk 8]
```diff
@@ -792,3 +792,219 @@ func (suite *KeeperTestSuite) _TestForEachStorage() {
...
+			suite.SetupTest()
```
Comment (zeroFruit) [inline x/evm/keeper/statedb_test.go:972]: Why `SetupTest()` for every case?

File: x/evm/keeper/statedb_test.go [hunk 9]
```diff
@@ -792,3 +792,219 @@ func (suite *KeeperTestSuite) _TestForEachStorage() {
...
+		name       string
+		initAmount int64
+		operation  func(stateDB *statedb.StateDB, from, to common.Address, amounts []int64)
+		/*
+			0: eth from before operation - eth from after operation
+			1: eth to after operation - eth to before operation
+			2: cosmos from before operation - cosmos from after operation
+			3: cosmos to after operation - cosmos to before operation
+		*/
+		expectedAmounts [4]int64
```
Comment (zeroFruit) [inline x/evm/keeper/statedb_test.go:816]: Question:
Is this test case testing the balance of two accounts in which the amount is differ depending on the flush/revert call?

File: x/evm/keeper/state_transition.go [hunk 10]
```diff
@@ -415,6 +417,7 @@ func (k *Keeper) ApplyMessageWithConfig(ctx sdk.Context, msg core.Message, trace
 		if err := stateDB.Commit(); err != nil {
 			return nil, sdkerrors.Wrap(err, "failed to commit stateDB")
 		}
+		write()
```
Comment (zeroFruit) [inline x/evm/keeper/state_transition.go:420]: `write()` is for committing the current branch of context on the Cosmos side?

File: x/evm/statedb/statedb.go [hunk 11]
```diff
@@ -431,22 +455,35 @@ func (s *StateDB) RevertToSnapshot(revid int) {
 	// Replay the journal to undo changes and remove invalidated snapshots
 	s.journal.revert(s, snapshot)
 	s.validRevisions = s.validRevisions[:idx]
+	s.cacheCtxs = s.cacheCtxs[:idx]
 }
 
 // Commit writes the dirty states to keeper
 // the StateDB object should be discarded after committed.
 func (s *StateDB) Commit() error {
+	if s.nextRevisionID == 0 {
```
Comment (zeroFruit) [inline x/evm/statedb/statedb.go:464]: What is `nextRevisionID` ?

File: x/evm/statedb/statedb.go [hunk 12]
```diff
@@ -431,22 +455,35 @@ func (s *StateDB) RevertToSnapshot(revid int) {
 	// Replay the journal to undo changes and remove invalidated snapshots
 	s.journal.revert(s, snapshot)
 	s.validRevisions = s.validRevisions[:idx]
+	s.cacheCtxs = s.cacheCtxs[:idx]
```
Comment (zeroFruit) [inline x/evm/statedb/statedb.go:458]: Is this code for syncing revert both of StateDB and Cosmos state?

**Review by hea9549 — APPROVED** [review]

**Comment by AgwaB** [comment]
It was verbally discussed with @zeroFruit

**Review by zeroFruit — APPROVED** [review]

**Pull request merged**